### ISSUE TITLE
[feature request] integrated clipboard manager

### ISSUE THREAD

**Issue by xi** [labwc/labwc#2042]
I find it deeply confusing that the clipboard is lost when I close an application (see #784). I know that the official solution is to setup a clipboard manager. But I think it would be so much simpler if the clipboard manager would be integrated in labwc itself.

**Comment by tokyo4j** [comment]
It's unlikely that we add builtin clipboard manager due to our policy.
But we'll need to document how to set up clipboard manager in https://labwc.github.io/integration.html.

**Comment by Consolatis** [comment]
I'd also don't want to accidentally store a bunch of 8k images in various formats including BMP in compositor memory (or whatever else some random application puts into the clipboard). Due to the wlr-data-data protocol supported by wlroots and by extension labwc an external clipboard manager for the people that want it should be enough.

Agree that we should better document setting something up.

**Comment by 01micko** [comment]
There are a few but I use `cliphist` that depends on `wl-clipboard` started in `autostart` like so:
```sh
type cliphist >/dev/null 2>&1 && wl-paste --watch cliphist store &
```
and for gui using `nwg-clipman` but not 100% happy with the setup though it does the job well enough.

As I use `sfwbar`,  I'm pretty sure someone was working on a clipboard widget. I'll do a bit more digging and _might_ offer up a PR for the integration page.

**Comment by stefonarch** [comment]
I first had a complicated setup I don't even remember and it had some downsides. Now using copyQ which has far to many options and features s for my needs but works perfectly, just using a shortcut for `copyq show`, its icon in lxqt-panel' systray  I've hidden.

**Comment by xi** [comment]
I did some more reading and this turned out to be much more complicated that I had expected. Maybe this was obvious for everyone but me. But I think it doesn't hurt to write it down.

TL;DR: clipboard managers are not a solution, they just make different trade offs. I believe labwc itself could make a more balanced trade off.

I first thought that the clipboard contains the actual bits that are being copied. But that is not the case. When you hit ctrl-c, no data is actually copied. Instead the data can be offered in multiple mime types at the same time, so that the receiving end can pick which mime type it prefers. For example, text copied from the web browser could be offered as both `text/html` and `text/plain`.

Of course, this mechanism only works as long as the application that is offering the data does not quit. After that, there is no one left to answer the request.

A common workaround for that problem is to use a clipboard manager like this:

```
wl-paste --watch cliphist store
```

In this case, `wl-paste` will call `cliphist store` whenever something is copied. It passes the copied data *in one of the available mime types*. `cliphist` will then replace the original offer with its own (`text/plain`). This means that the clipboard contents will not be lost when the original source goes away. But it also has some drawbacks:

- Most mime type information is lost
- The data needs to be transferred even if it is never actually pasted

[clipmon](https://git.sr.ht/~whynothugo/clipmon) is another clipboard manager that makes slightly different trade offs: It also replaces the original offer with its own, but it preserves all mime types. This solves the first drawback, but amplifies the second one. For example, the same image might have to be stored in multiple different formats.

So in summary, the following goals are in conflict:

- offer content in different mime types
- keep the clipboard alive when an application quits
- avoid unnecessary data transfer and storage

It would be nice if we could keep the original offer intact, and only replace it with a fallback if the application actually quits. The fallback could be restricted (e.g. like the one provided by `cliphist`) to avoid wasting resources. Still, I think this would be a more balanced solution than what is currently available.

I don't think this can be done by an external clipboard manager though. The compositor should have all the information that is needed, so I am back to proposing to integrate this into labwc itself.

**Comment by droc12345** [comment]
IMO, something like this is not part of what a compositor should do.
Maybe a DE like gnome, kde, etc might implement this as part of their suite of apps

**Comment by Consolatis** [comment]
> It would be nice if we could keep the original offer intact, and only replace it with a fallback if the application actually quits. The fallback could be restricted (e.g. like the one provided by `cliphist`) to avoid wasting resources. Still, I think this would be a more balanced solution than what is currently available.
> 
> I don't think this can be done by an external clipboard manager though. The compositor should have all the information that is needed, so I am back to proposing to integrate this into labwc itself.

I didn't verify this but I think this could be implemented by an external client:
- Using the privileged protocol you get notified about a new [selection](https://gitlab.freedesktop.org/wlroots/wlr-protocols/-/blob/master/unstable/wlr-data-control-unstable-v1.xml?ref_type=heads#L116) and then request text/plain and store the result somewhere.
- Once there is a new selection for text/plain it is requested again and the old content get overwritten.
- Once the the selection is set to NULL an own text/plain offer is created with the stored data

**Comment by johanmalm** [comment]
@xi Thanks for researching.

I prefer to use clients where possible. It is not obvious to me what advantage the compositor has over clients in this context. I get your general point that "the compositor should have all the information", but what specifically is missing for a client to be able to do this?

**Comment by xi** [comment]
Thanks for the responses. I had expected that it would be hard to know when an application quits. But it turns out the data control protocol doesn't make that too hard. I got the behavior that I described with the following script:

```sh
#!/bin/sh -e
path="$XDG_RUNTIME_DIR/clipboard.txt"

if wl-paste -l >/dev/null 2>&1; then
	input=$(cat)
	if [ -n "$input" ]; then
		printf '%s' "$input" > "$path"
	fi
elif [ -f "$path" ]; then
	wl-copy < "$path"
fi
```

Now you have convinced my that this *can* be done in an external clipboard manager, what should we recommend in the documentation? Should we outline the different trade offs (which could be quote long)? Should we link to some example clipboard managers and hope that their documentation speaks for itself (which I doubt)?

This is ultimately not an issue with labwc but with the ecosystem. But for users this does look like a labwc bug. So I think we need an answer.

**Comment by johanmalm** [comment]
> Now you have convinced my that this _can_ be done in an external clipboard manager, what should we recommend in the documentation? Should we outline the different trade offs (which could be quote long)? Should we link to some example clipboard managers and hope that their documentation speaks for itself (which I doubt)?

I suspect many users would be grateful for an understanding of the options relating to this.

I'm happy to have a dedicated page on the website on this. We can link to it from `integration` or similar.

**Comment by WhyNotHugo** [comment]
I wrote [clipmon](https://git.sr.ht/~whynothugo/clipmon) and a few other prototype clipboard managers that all have big caveats. Hopefully I can avoid others repeating the same [failed] experiments.

Using `wlr-data-control-unstable-v1` for this purpose is racey; when the clipboard is taken multiple times in quick succession, the clipboard manager has no way to assert whether the clipboard was taken by another client or itself.

A common approach to retain the clipboard when applications exit is for the clipboard manager to immediately take any selections and keep a copy of all data itself. The main problem with this is that if I copy an image in an image editor, the clipboard manager immediately reads all clipboard data in all mime types, which requires the image editor to encode the image every supported formats each time that I press <kbd>Ctrl</kbd>+<kbd>c</kbd>. If I copy a portion of an image file in GIMP, my CPU suddenly spikes as GIMP starts encoding the image in 18 different formats for the clipboard manager. This is useful if I'll later paste on another application, but useless work in cases when I'm about to paste into GIMP.

A clipboard manager has no way of knowing if a copy operation will be followed by an intra-application paste, no paste, or inter-application paste. It must always fetch all mime types, which requires the copying application to encode into all mime types.

Several clipboard manager work around this issue by copying only `text/plain` data. This actually breaks the clipboard, because when I copy a file, the selection is advertised as several mime types including `text/uri-list` and `text/plain`. Clipboard manager that ignore everything except `text/plain` break copying files, because they discard the mime-type that is important in this case. And, of course, they still don't fix the underlying issue for anything that is not plain text.

If you only use the clipboard for plain text and absolutely nothing else, this last approach might kind of work for you. If, like me, you copy a lot of images and files and other sorts of data, then they won't.

---

I don't necessarily think that implementing clipboard persistence _inside_ the compositor is a great idea. But with the current Wayland protocols, it is not possible for a client reliably implement this, and there is a very high runtime cost to the existing approaches.

Some ideas have not been tried/explored so far:

- Have a dedicated protocol for a client to inform when they are about to exit and need their selection persisted.
- Simply have clients linger in the background if they own a selection. This is a very clean solution (and only fails with applications that crash), but requires modifying almost every existing GUI applications out there.

**Comment by Consolatis** [comment]
> Using `wlr-data-control-unstable-v1` for this purpose is racey; when the clipboard is taken multiple times in quick succession, the clipboard manager has no way to assert whether the clipboard was taken by another client or itself.

The clipboard manager could expose an own clipboard manager internal mime type to combat this. So it would be able to detect its own selection that way. It is pretty hacking but I don't see why that wouldn't work.

However for the narrow case discussed here from what I understand (keeping text/plain and maybe text/uri-list selection if client exits) fetching both types and storing it in ram but only setting them when the selection is set to `NULL` should be enough.

**Comment by WhyNotHugo** [comment]
> The clipboard manager could expose an own clipboard manager internal mime type to combat this. So it would be able to detect its own selection that way. It is pretty hacky but I don't see why that wouldn't work.

I'll try that out. It _should_ work for that particular race.

**Comment by hermitguo** [comment]
ENV: ubuntu budgie 26.04 wayland (labwc ) + fctix5 + oracle virtual box 7.2.12(win11)

 I'm experiencing issues copying terminal text from the host machine to Windows 11 on VirtualBox 7.2.12—sometimes it works, sometimes not; sometimes I can copy entire sections but not individual lines.

**host (Terminal) --> oracle virtual box - win11 (office 2024) : fail**


<img width="808" height="539" alt="Image" src="https://github.com/user-attachments/assets/ce7742a3-e309-4ac3-9240-2a3b2390f892" />


The text copied from the host machine specifically for OnlyOffice can be pasted onto Windows 11 on VirtualBox 7.2.12 running Microsoft Office 2024 under various conditions.

**host (onlyoffice 9) --> oracle virtual box - win11 (office 2024) : success**

This bug may be specific to LabWC or Budgie 26.04; Budgie 24.04 X11 does not exhibit this issue.


```
ubuntu@master:~/.config/autostart$ labwc --version
labwc 0.9.3 (+xwayland +nls +rsvg +libsfdo)
ubuntu@master:~/.config/autostart$
```

**Comment by johanmalm** [comment]
IIRC there is one way that does not work (XWayland to native or the other way around), cannot remember which. 

We ought to document this and describe why.

Help needed to remember the technical reasoning.
